pkg/reconciler/topology/partitionset/partitionset_reconcile.go

```diff
@@ -0,0 +1,208 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package partitionset
+
+import (
+	"context"
+	"sort"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	corev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	topologyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/topology/v1alpha1"
+)
+
+func (c *controller) reconcile(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	labelSelector := labels.Everything()
+	if partitionSet.Spec.ShardSelector == nil {
+		var err error
+		labelSelector, err = metav1.LabelSelectorAsSelector(partitionSet.Spec.ShardSelector)
+		// This should not happen due to OpenAPI validation
+		if err != nil {
+			conditions.MarkFalse(
+				partitionSet,
+				topologyv1alpha1.PartitionSetValid,
+				topologyv1alpha1.PartitionSetInvalidSelectorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				err.Error(),
+			)
+			conditions.MarkFalse(
+				partitionSet,
+				topologyv1alpha1.PartitionsReady,
+				topologyv1alpha1.ErrorGeneratingPartitionsReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"",
+			)
+			// No need to requeue if the selector is not valid
+			return nil
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionSetValid)
+
+	shards, err := c.listShards(labelSelector)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error listing Shards",
+		)
+		return err
+	}
+
+	oldPartitions, err := c.getPartitionsByPartitionSet(partitionSet)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return err
+	}
+
+	var matchLabelsMap map[string]map[string]string
+	if partitionSet.Spec.ShardSelector != nil {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, partitionSet.Spec.ShardSelector.MatchLabels)
+	} else {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, nil)
+	}
+	partitionSet.Status.Count = uint16(len(matchLabelsMap))
+	existingMatches := map[string]struct{}{}
```

Comment:
this can be made type-safe. Compare https://goplay.space/#IxSJ1_iWaY9.

Reply:
I meant the key, i.e. that the key is a string but could be a struct. No need to think about representation.